include/tudocomp/coders/AdaptiveHuffmanCoder.hpp

```diff
@@ -0,0 +1,857 @@
+            while (correspondingLeaf != nullptr) {
+                slide_and_increment(&codingtree->root, &correspondingLeaf);
+
+                implicit_numbering(codingtree);
+            }
+            if (leafToIncrement != nullptr) {
+                slide_and_increment(&codingtree->root, &leafToIncrement);
+
+                implicit_numbering(codingtree);
+            }
+
+        }
+
+
+
+        /**
+         * Encode
+         */
+
+        /**
+         * TODO
+         * \brief Writes the code for a given symbol into the queue code_write
+         * @param do_code
+         * @param code_write
+         * @param symbol
+         * @param pTree
+         */
+        inline void
+        gen_code(std::queue<bool> &do_code, std::queue<bool> &code_write, const uliteral_t symbol, node *pTree) {
+
+            // symbol found?
+            if ((pTree)->symbol == symbol && (pTree)->is_leaf() && (pTree)->weight != 0) {
+                while (!do_code.empty()) {
+                    bool next_bit = do_code.front();
+                    do_code.pop();
+                    code_write.push(next_bit);
+                }
+                return;
+            }
+
+            std::queue<bool> temp_left(do_code);
+            if ((pTree)->has_left()) {
+                temp_left.push(false);
+                gen_code(temp_left, code_write, symbol, (pTree)->left);
+            }
+
+            std::queue<bool> temp_right(do_code);
+            if ((pTree)->has_right()) {
+                temp_right.push(true);
+                gen_code(temp_right, code_write, symbol, (pTree)->right);
+            }
+
+            return;
+        }
+
+        /**
+         * \brief Creates the nyt-code for the current tree
+         * @param do_code Temporary queue
+         * @param write_code Result queue
+         * @param pTree Current node
+         */
+        inline void gen_nyt_code(std::queue<bool> &do_code, std::queue<bool> &write_code, node *pTree) {
+
+            if ((pTree)->weight == 0 && (pTree)->is_leaf()) {
+                while (!do_code.empty()) {
+                    bool next_bit = do_code.front();
+                    do_code.pop();
+                    write_code.push(next_bit);
+                }
+            }
+
+            std::queue<bool> temp_left(do_code);
+            if ((pTree)->has_left()) {
+                temp_left.push(false);
+                gen_nyt_code(temp_left, write_code, (pTree)->left);
+            }
+
+            std::queue<bool> temp_right(do_code);
+            if ((pTree)->has_right()) {
+                temp_right.push(true);
+                gen_nyt_code(temp_right, write_code, (pTree)->right);
+            }
+
+        }
+
+        /**
+         * Creates the standart encoding for a symbol
+         * @param code
+         * @param symbol
+         */
+        inline void gen_standard_code(std::queue<bool> &code, uliteral_t symbol) {
+            unsigned int symbol_size = sizeof(uliteral_t)*CHAR_BIT;
```

Comment:
Was about to say that we have a `ULITERAL_BITS`, but turns out we don't. 😅 (I see this as a bug on our end...)